### PR TITLE
feat(emqx_ws_connection): #4184 check http header to know real IP/port

### LINKED ISSUE #4184: listener.ws.external.proxy_address_header and listener.ws.external.proxy_port_header not working
**Environment**:

- EMQ X version: 4.2.7 Official Docker container

**What happened and what you expected to happen**:

I configured emqx according to the last user guide wit the following env variables in my docker-compose file:
      - EMQX_listener__ws__external__proxy_address_header=X-Forwarded-For
      - EMQX_listener__ws__external__proxy_port_header=X-Forwarded-Port

nginx is also configured according to the guide as follows:

        proxy_set_header X-Forwarded-Port $server_port;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;

but I still have the nginx private IP address in the client list.

**How to reproduce it (as minimally and precisely as possible)**:

Configure emqx 4.2.7 broker with:
- EMQX_listener__ws__external__proxy_address_header=X-Forwarded-For
- EMQX_listener__ws__external__proxy_port_header=X-Forwarded-Port

and nginx with:

proxy_set_header X-Forwarded-Port $server_port;
proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;

### PR DESCRIPTION
**Pull request by z8674558**
Fixes #4184

**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/emqx/emqx/blob/master/CONTRIBUTING.md.

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)
- [ ] In case of non-backward compatible changes, reviewer should check this item as a write-off, and add details in **Backward Compatibility** section

## More information

### CODE REVIEW

File: priv/emqx.schema [hunk 1]
```diff
@@ -1531,6 +1531,14 @@ end}.
   {datatype, string}
 ]}.
 
+{mapping, "listener.ws.$name.proxy_address_header", "emqx.listeners", [
+  {datatype, {enum, [x_forwarded_for]}}
```
Comment (zmstone) [inline priv/emqx.schema:1535]: should be a string type,
as a user, I would not like to edit schema if I want to use a different header name.

Comment (zmstone) [reply priv/emqx.schema:1535]: Add default value "X-Forwarded-For"

File: etc/emqx.conf [hunk 2]
```diff
@@ -1608,6 +1608,20 @@ listener.ws.external.access.1 = allow all
 ## Default: mqtt, mqtt-v3, mqtt-v3.1.1, mqtt-v5
 ## listener.ws.external.supported_subprotocols = mqtt, mqtt-v3, mqtt-v3.1.1, mqtt-v5
 
+## Use X-Forwarded-For header for real source IP if the EMQ X cluster is
+## deployed behind NGINX or HAProxy.
+##
+## Value: Enum
```
Comment (zmstone) [inline etc/emqx.conf:1614]: should be a string type.

File: etc/emqx.conf [hunk 3]
```diff
@@ -1608,6 +1608,20 @@ listener.ws.external.access.1 = allow all
 ## Default: mqtt, mqtt-v3, mqtt-v3.1.1, mqtt-v5
 ## listener.ws.external.supported_subprotocols = mqtt, mqtt-v3, mqtt-v3.1.1, mqtt-v5
 
+## Use X-Forwarded-For header for real source IP if the EMQ X cluster is
```
Comment (zmstone) [inline etc/emqx.conf:1611]:
```suggestion
## Specify which HTTP header for real source IP if the EMQ X cluster is
```

File: src/emqx_ws_connection.erl [hunk 4]
```diff
@@ -253,7 +253,33 @@ websocket_init([Req, Opts]) ->
                    #{src_address := SrcAddr, src_port := SrcPort} ->
                        {SrcAddr, SrcPort};
                    _ ->
-                       cowboy_req:peer(Req)
+                       {PeerAddr, PeerPort} = cowboy_req:peer(Req),
+                        Addr = case proplists:get_value(proxy_address_header, Opts) of
```
Comment (zmstone) [inline src/emqx_ws_connection.erl:257]: break this out to a small function ?

File: src/emqx_ws_connection.erl [hunk 5]
```diff
@@ -253,7 +253,33 @@ websocket_init([Req, Opts]) ->
...
+                            x_forwarded_for ->
```
Comment (zmstone) [inline src/emqx_ws_connection.erl:258]: should get back a string, `HeaderName` then call 
`cowboy_req:header(lowercase(HeaderName), Req, <<>>)`

File: src/emqx_ws_connection.erl [hunk 6]
```diff
@@ -253,7 +253,33 @@ websocket_init([Req, Opts]) ->
...
+                                XForwardedFor = cowboy_req:header(<<"x-forwarded-for">>, Req, <<>>),
+                                ClientAddr = hd(string:tokens(binary_to_list(XForwardedFor), ", ")),
+                                case inet:parse_address(ClientAddr) of
+                                    {ok, A} ->
+                                        A;
+                                    _ ->
+                                        PeerAddr
+                                end;
+                            _ ->
```
Comment (zmstone) [inline src/emqx_ws_connection.erl:267]: if the schema is created with a default value, this clause is not necessary.

**Review by zmstone — CHANGES_REQUESTED** [review]

File: priv/emqx.schema [hunk 7]
```diff
@@ -1715,6 +1725,16 @@ end}.
   {datatype, string}
 ]}.
 
+{mapping, "listener.wss.$name.proxy_address_header", "emqx.listeners", [
+  {default, "X-Forwarded-For"},
```
Comment (HJianBo) [inline priv/emqx.schema:1729]: The default value here means that the behavior of parsing `X-Forward-*` is turned on by default, is this purposely done here?

My opinion is that, first of all, this is different from the previous design (in 2.3.11). The previous logic required the configuration to be manually enabled to allow parsing of `X-Forward-*`; when the configuration was commented out, it meant that `X-Forward-*` would not be parsed.

Comment (HJianBo) [reply priv/emqx.schema:1729]: By default we should turn this off, is this ok? 

Or at least it should allow the user a way to turn it off.

**Comment by zmstone** [comment]
@HJianBo I don’t think it’s necessary to turn it off by default.
Showing haproxy ip as client ip by default is a bug .
And http layer headers should override tcp layer’s binary PROXY protocol ( maybe turn on binary protocol flag by default too)

X-Forwarded-For is almost the standard, see https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For

**Review by HJianBo — APPROVED** [review]

**Review by zmstone — APPROVED** [review]

**Pull request merged**